src/github.com/matrix-org/dendrite/roomserver/api/input.go

```diff
@@ -36,7 +40,51 @@ type InputRoomEvent struct {
 	// For example many matrix events forget to reference the m.room.create event even though it is needed for auth.
 	// (since synapse allows this to happen we have to allow it as well.)
 	AuthEventIDs []string
+	// Whether the state is supplied as a list of event IDs or whether it
+	// should be derived from the state at the previous events.
+	HasState bool
 	// Optional list of state event IDs forming the state before this event.
 	// These state events must have already been persisted.
 	StateEventIDs []string
 }
+
+// UnmarshalJSON implements json.Unmarshaller
+func (ire *InputRoomEvent) UnmarshalJSON(data []byte) error {
```

Comment:
Why are you implementing your own functions here? `UnmarshalJSON` seems equivalent to just using the builtin?